eth/enr.go

```diff
@@ -0,0 +1,245 @@
+		//      remote FORK_NEXT matches with the locally following fork block number,
+		//      connect.
+		//        Remote node is currently syncing. It might eventually diverge from
+		//        us, but at this current point in time we don't have enough information.
+		//   3. If the remote FORK_CSUM is a superset of the local past forks and can
+		//      be completed with locally known future forks, connect.
+		//        Local node is currently syncing. It might eventually diverge from
+		//        the remote, but at this current point in time we don't have enough
+		//        information.
+		//   4. Reject in all other cases.
+		var (
+			head       = headfn()
+			remoteSum  = binary.BigEndian.Uint32(entry[4:8])
+			remoteNext = binary.BigEndian.Uint32(entry[8:12])
+		)
+		for i, fork := range forks {
+			// If our head is beyond this fork, continue to the next (we have a dummy
+			// fork of maxuint32 as the last item to always fail this check eventually).
+			if head > fork {
+				continue
+			}
+			// Found the first unpassed fork block, check if our current state matches
+			// the remote checksum (rule #1).
+			if sums[i] == remoteSum {
+				// Yay, fork checksum matched, ignore any upcoming fork
+				return nil
+			}
+			// The local and remote nodes are in different forks currently, check if the
+			// remote checksum is a subset of our local forks (rule #2).
+			for j := 0; j < i; j++ {
+				if sums[j] == remoteSum {
+					// Remote checksum is a subset, validate based on the announced next fork
+					if uint32(forks[j]) != remoteNext {
+						return errENRRemoteStale
+					}
+					return nil
+				}
+			}
+			// Remote chain is not a subset of our local one, check if it's a superset by
+			// any chance, signalling that we're simply out of sync (rule #3).
+			for j := i + 1; j < len(sums); j++ {
+				if sums[j] == remoteSum {
+					// Yay, remote checksum is a superset, ignore upcoming forks
+					return nil
+				}
+			}
+			// No exact, subset or superset match. We are on differing chains, reject.
+			return errENRLocalStale
+		}
+		log.Error("Impossible eth ENR validation", "record", fmt.Sprintf("%x", entry))
+		return nil // Something's very wrong, accept rather than reject
+	}
+}
+
+// makeGenesisChecksum calculates the ENR checksum for the genesis block.
+func makeGenesisChecksum(hash common.Hash) uint32 {
+	var checksum uint32
+	for i := 0; i < 8; i++ {
+		for j := 0; j < 4; j++ {
+			checksum ^= uint32(hash[i*4+j]) << uint32(24-8*j)
+		}
+	}
+	return checksum
+}
+
+// gatherForks gathers all the known forks and creates a sorted list out of them.
+func gatherForks(config *params.ChainConfig) []uint64 {
+	// Gather all the fork block numbers via reflection
+	kind := reflect.TypeOf(params.ChainConfig{})
+	conf := reflect.ValueOf(config).Elem()
+
+	var forks []uint64
+	for i := 0; i < kind.NumField(); i++ {
+		// Fetch the next field and skip non-fork rules
+		field := kind.Field(i)
+		if !strings.HasSuffix(field.Name, "Block") {
```

Comment:
This is a bit of a hack. Wouldn't it be better to explicity list the forks instead? But otoh, I guess you want this to 'just work' in the future too... Not sure what the best way is.